accepted/2021/statics-in-interfaces/README.md

```diff
@@ -6,8 +6,1267 @@
+        // returning nint/nuint to better support big data and its possible we would want to break compat here for a better experience
+
+        long GetBitLength();
+
+        int GetByteCount(bool isUnsigned);
+
+        byte[] ToByteArray()
+        byte[] ToByteArray(bool isUnsigned = false, bool isBigEndian = false);
+
+        bool TryWriteBytes(Span<byte> destination, out int bytesWritten);
+        bool TryWriteBytes(Span<byte> destination, out int bytesWritten, bool isUnsigned, bool isBigEndian);
+    }
+
+    public interface IBinaryInteger<TSelf> : IBinaryNumber<TSelf>
+        where TSelf : IBinaryInteger<TSelf>
+    {
+        // We might want to support "big multiplication" for fixed width types
+        // This would return a tuple (TSelf high, TSelf low) or similar, to match what we do for System.Math
+
+        // Returning int is currently what BitOperations does, however this can be cumbersome or prohibitive
+        // in various algorithms where returning TSelf is better.
+
+        static abstract int LeadingZeroCount(TSelf value);
+
+        static abstract int PopCount(TSelf value);
+
+        static abstract int TrailingZeroCount(TSelf value);
+
+        // Much like with shift, the rotateAmount should likely support being TSelf and potentially other types
+
+        static abstract TSelf RotateLeft(TSelf value, int rotateAmount);
+
+        static abstract TSelf RotateRight(TSelf value, int rotateAmount);
+    }
+
+    public interface IFloatingPoint<TSelf>
+        : ISignedNumber<TSelf>
+        where TSelf : IFloatingPoint<TSelf>
+    {
+        // This currently implies IEEE floating-point types and so decimal does not implement it
+        // If we want decimal to implement it, we would need to move several methods down and define
+        // some IeeeFloatingPoint interface instead. This wasn't done for simplicity and because decimal
+        // is not an industry standard type.
+
+        // TODO: The Exponent and Signifcand need to be exposed here
+        // They should return INumber<TSelf> here and IBinaryNumber<TSelf> on IBinaryFloatingPoint<TSelf>
+
+        // TODO: We may want to expose constants related to the subnormal and finite boundaries
+
+        // The following constants are defined by System.Math and System.MathF today
+
+        static abstract TSelf E { get; }
+
+        static abstract TSelf PI { get; }
+
+        static abstract TSelf Tau { get; }
+
+        // The following methods are defined by System.Math and System.MathF today
+        // Exposing them on the interfaces means they will become available as statics on the primitive types
+        // This, in a way, obsoletes Math/MathF and brings float/double inline with how non-primitive types support similar functionality
+        // API review will need to determine if we'll want to continue adding APIs to Math/MathF in the future
+
+        static abstract TSelf Acos(TSelf x);
```

Comment:
Why are many of these static rather than instance?  i.e. I'm not sure why this isn't just `TSelf Acos();`?

Reply:
Because these kinds of operations have traditionally been `static` in .NET.

Also, there is a perf consideration here because `this` is `byref` (or `inref` for `readonly struct`) which can hurt inlining and other optimizations done by the JIT.